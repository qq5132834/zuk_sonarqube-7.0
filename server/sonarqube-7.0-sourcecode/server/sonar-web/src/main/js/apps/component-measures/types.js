/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/*:: import type { Measure, MeasureEnhanced } from '../../components/measure/types'; */

/*:: type ComponentIntern = {
  isFavorite?: boolean,
  isRecentlyBrowsed?: boolean,
  key: string,
  match?: string,
  name: string,
  organization?: string,
  project?: string,
  qualifier: string
}; */

/*:: export type Component = ComponentIntern & { measures?: Array<Measure> }; */

/*:: export type ComponentEnhanced = ComponentIntern & {
  value?: ?string,
  leak?: ?string,
  measures: Array<MeasureEnhanced>
}; */

/*:: export type Paging = {
  pageIndex: number,
  pageSize: number,
  total: number
}; */

/*:: export type Period = {
  index: number,
  date: string,
  mode: string,
  parameter?: string
}; */

/*:: export type Query = {
  metric: ?string,
  selected: ?string,
  view: string
}; */
