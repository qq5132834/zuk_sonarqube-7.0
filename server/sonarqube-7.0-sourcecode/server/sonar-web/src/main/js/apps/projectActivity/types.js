/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow

/*::
export type Event = {
  key: string,
  name: string,
  category: string,
  description?: string
};
*/

/*::
export type Analysis = {
  key: string,
  date: Date,
  events: Array<Event>
};
*/

/*::
export type HistoryItem = { date: Date, value: string };
*/

/*::
export type MeasureHistory = { metric: string, history: Array<HistoryItem> };
*/

/*::
export type Metric = {
  custom?: boolean,
  hidden?: boolean,
  key: string,
  name: string,
  type: string
};
*/

/*::
export type Paging = {
  pageIndex: number,
  pageSize: number,
  total: number
};
*/

/*::
export type Query = {
  category: string,
  customMetrics: Array<string>,
  from?: Date,
  graph: string,
  project: string,
  to?: Date,
  selectedDate?: Date
};
*/
