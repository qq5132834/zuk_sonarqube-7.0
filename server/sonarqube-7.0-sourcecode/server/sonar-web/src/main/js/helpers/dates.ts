/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  differenceInDays as _differenceInDays,
  differenceInSeconds as _differenceInSeconds,
  differenceInYears as _differenceInYears,
  isSameDay as _isSameDay,
  parse,
  startOfDay as _startOfDay
} from 'date-fns';

function pad(number: number) {
  if (number < 10) {
    return '0' + number;
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function startOfDay(date: Date): Date {
  return _startOfDay(date);
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
  return _isSameDay(dateLeft, dateRight);
}

export function differenceInYears(dateLeft: Date, dateRight: Date): number {
  return _differenceInYears(dateLeft, dateRight);
}

export function differenceInDays(dateLeft: Date, dateRight: Date): number {
  return _differenceInDays(dateLeft, dateRight);
}

export function differenceInSeconds(dateLeft: Date, dateRight: Date): number {
  return _differenceInSeconds(dateLeft, dateRight);
}
