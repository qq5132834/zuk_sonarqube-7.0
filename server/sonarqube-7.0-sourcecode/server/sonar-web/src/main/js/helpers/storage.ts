/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const PROJECTS_DEFAULT_FILTER = 'sonarqube.projects.default';
const PROJECTS_FAVORITE = 'favorite';
const PROJECTS_ALL = 'all';

const PROJECTS_VIEW = 'sonarqube.projects.view';
const PROJECTS_VISUALIZATION = 'sonarqube.projects.visualization';
const PROJECTS_SORT = 'sonarqube.projects.sort';

const PROJECT_ACTIVITY_GRAPH = 'sonarqube.project_activity.graph';
const PROJECT_ACTIVITY_GRAPH_CUSTOM = 'sonarqube.project_activity.graph.custom';

function save(key: string, value?: string): void {
  try {
    if (value) {
      window.localStorage.setItem(key, value);
    } else {
      window.localStorage.removeItem(key);
    }
  } catch (e) {
    // usually that means the storage is full
    // just do nothing in this case
  }
}

export function saveFavorite(): void {
  save(PROJECTS_DEFAULT_FILTER, PROJECTS_FAVORITE);
}

export function isFavoriteSet(): boolean {
  const setting = window.localStorage.getItem(PROJECTS_DEFAULT_FILTER);
  return setting === PROJECTS_FAVORITE;
}

export function saveAll(): void {
  save(PROJECTS_DEFAULT_FILTER, PROJECTS_ALL);
}

export function isAllSet(): boolean {
  const setting = window.localStorage.getItem(PROJECTS_DEFAULT_FILTER);
  return setting === PROJECTS_ALL;
}

export function saveView(view?: string): void {
  save(PROJECTS_VIEW, view);
}

export function getView(): string | null {
  return window.localStorage.getItem(PROJECTS_VIEW);
}

export function saveVisualization(visualization?: string): void {
  save(PROJECTS_VISUALIZATION, visualization);
}

export function getVisualization(): string | null {
  return window.localStorage.getItem(PROJECTS_VISUALIZATION);
}

export function saveSort(sort?: string): void {
  save(PROJECTS_SORT, sort);
}

export function getSort(): string | null {
  return window.localStorage.getItem(PROJECTS_SORT);
}

export function saveCustomGraph(metrics?: string[]): void {
  save(PROJECT_ACTIVITY_GRAPH_CUSTOM, metrics ? metrics.join(',') : '');
}

export function getCustomGraph(): string[] {
  const customGraphs = window.localStorage.getItem(PROJECT_ACTIVITY_GRAPH_CUSTOM);
  return customGraphs ? customGraphs.split(',') : [];
}

export function saveGraph(graph?: string): void {
  save(PROJECT_ACTIVITY_GRAPH, graph);
}

export function getGraph(): string {
  return window.localStorage.getItem(PROJECT_ACTIVITY_GRAPH) || 'issues';
}
