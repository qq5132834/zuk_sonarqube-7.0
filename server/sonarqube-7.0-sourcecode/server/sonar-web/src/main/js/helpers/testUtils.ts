/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  element.simulate('click', { ...mockEvent, ...event });
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper, value: string, event = {}): void {
  element.simulate('change', {
    target: { value },
    currentTarget: { value },
    ...event
  });
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl(node: React.ReactElement<any>, options: ShallowRendererProps = {}) {
  return shallow(node, { ...options, context: { intl, ...options.context } });
}
