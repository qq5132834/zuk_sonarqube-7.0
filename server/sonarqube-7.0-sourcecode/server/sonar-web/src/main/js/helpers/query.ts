/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isNil, omitBy } from 'lodash';
import { isValidDate, parseDate, toNotSoISOString } from './dates';

export interface RawQuery {
  [x: string]: any;
}

function arraysEqual(a: RawQuery, b: RawQuery): boolean {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

export function queriesEqual(a: RawQuery, b: RawQuery): boolean {
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);

  if (keysA.length !== keysB.length) {
    return false;
  }

  return keysA.every(
    key =>
      Array.isArray(a[key]) && Array.isArray(b[key])
        ? arraysEqual(a[key], b[key])
        : a[key] === b[key]
  );
}

export function cleanQuery(query: RawQuery): RawQuery {
  return omitBy(query, isNil);
}

export function parseAsBoolean(value: string | undefined, defaultValue: boolean = true): boolean {
  return value === 'false' ? false : value === 'true' ? true : defaultValue;
}

export function parseAsDate(value?: string): Date | undefined {
  if (value) {
    const date = parseDate(value);
    if (isValidDate(date)) {
      return date;
    }
  }
  return undefined;
}

export function parseAsFacetMode(facetMode: string): string {
  return facetMode === 'debt' || facetMode === 'effort' ? 'effort' : 'count';
}

export function parseAsString(value: string | undefined): string {
  return value || '';
}

export function parseAsArray(
  value: string | undefined,
  itemParser: (x: string) => string
): string[] {
  return value ? value.split(',').map(itemParser) : [];
}

export function serializeDate(value?: Date): string | undefined {
  if (value != null && value.toISOString) {
    return toNotSoISOString(value);
  }
  return undefined;
}

export function serializeString(value: string): string | undefined {
  return value || undefined;
}

export function serializeStringArray(value: string[] | undefined[]): string | undefined {
  return value && value.length ? value.join() : undefined;
}
