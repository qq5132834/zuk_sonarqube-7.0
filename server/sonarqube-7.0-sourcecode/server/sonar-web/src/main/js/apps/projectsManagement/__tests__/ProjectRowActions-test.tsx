/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ProjectRowActions, { Props } from '../ProjectRowActions';
import { Visibility } from '../../../app/types';
import { click } from '../../../helpers/testUtils';

jest.mock('../../../api/components', () => ({
  getComponentShow: jest.fn(() => Promise.reject(undefined))
}));

jest.mock('../../../api/nav', () => ({
  getComponentNavigation: jest.fn(() => Promise.resolve())
}));

const project = {
  id: '',
  key: 'project',
  name: 'Project',
  organization: 'org',
  qualifier: 'TRK',
  visibility: Visibility.Private
};

it('restores access', async () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();

  wrapper.prop<Function>('onToggleClick')();
  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('.js-restore-access'));
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('applies permission template', () => {
  const onApplyTemplate = jest.fn();
  const wrapper = shallowRender({ onApplyTemplate });
  click(wrapper.find('.js-apply-template'));
  expect(onApplyTemplate).toBeCalledWith(project);
});

function shallowRender(props: Partial<Props> = {}) {
  const wrapper = shallow(
    <ProjectRowActions
      currentUser={{ login: 'admin' }}
      onApplyTemplate={jest.fn()}
      project={project}
      {...props}
    />
  );
  (wrapper.instance() as ProjectRowActions).mounted = true;
  return wrapper;
}
