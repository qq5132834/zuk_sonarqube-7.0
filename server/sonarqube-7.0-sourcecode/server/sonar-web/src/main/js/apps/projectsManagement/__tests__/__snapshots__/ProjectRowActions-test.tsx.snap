// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restores access 1`] = `
<ActionsDropdown
  onToggleClick={[Function]}
>
  <ActionsDropdownItem
    className="js-apply-template"
    onClick={[Function]}
  >
    projects_role.apply_template
  </ActionsDropdownItem>
</ActionsDropdown>
`;

exports[`restores access 2`] = `
<ActionsDropdown
  onToggleClick={[Function]}
>
  <ActionsDropdownItem
    className="js-restore-access"
    onClick={[Function]}
  >
    global_permissions.restore_access
  </ActionsDropdownItem>
  <ActionsDropdownItem
    className="js-apply-template"
    onClick={[Function]}
  >
    projects_role.apply_template
  </ActionsDropdownItem>
</ActionsDropdown>
`;

exports[`restores access 3`] = `
<ActionsDropdown
  onToggleClick={[Function]}
>
  <ActionsDropdownItem
    className="js-restore-access"
    onClick={[Function]}
  >
    global_permissions.restore_access
  </ActionsDropdownItem>
  <ActionsDropdownItem
    className="js-apply-template"
    onClick={[Function]}
  >
    projects_role.apply_template
  </ActionsDropdownItem>
  <RestoreAccessModal
    currentUser={
      Object {
        "login": "admin",
      }
    }
    onClose={[Function]}
    onRestoreAccess={[Function]}
    project={
      Object {
        "id": "",
        "key": "project",
        "name": "Project",
        "organization": "org",
        "qualifier": "TRK",
        "visibility": "private",
      }
    }
  />
</ActionsDropdown>
`;
